File: one_fm/templates/pages/applicant-docs.py
import os, io
from google.cloud import vision
import frappe
from frappe.utils import cstr
import json
import frappe.sessions
import base64
import datetime
import hashlib

from dateutil.parser import parse


@frappe.whitelist(allow_guest=True)
def populate_nationality():
    return frappe.get_list('Nationality', pluck='name')

@frappe.whitelist(allow_guest=True)
def fetch_nationality(code):
    country = frappe.get_value('Country', {'code_alpha3':code},["country_name"])
    return frappe.get_value('Nationality', {'country':country},["name"])


@frappe.whitelist(allow_guest=True)
def token():
    return frappe.local.session.data.csrf_token

@frappe.whitelist(allow_guest=True)
def get_civil_id_text():
    """This API redirects the image fetched from frontend and 
    runs it though Google Vision API, each side at a time.

    Args:
        images (json): Consist of two base64 encoded strings(front_side, back_side).
        is_kuwaiti (int): 0 for non-Kuwaiti and 1 for Kuwaiti

    Returns:
        result: dictionary consisting of text fetched from both front and back side.
    """    
    try:
        
        result = {}
        
        #initialize google vision client library
        os.environ['GOOGLE_APPLICATION_CREDENTIALS'] = cstr(frappe.local.site) + frappe.local.conf.google_application_credentials
        client = vision.ImageAnnotatorClient()
       
        front_civil = frappe.local.form_dict['front_civil']
        back_civil = frappe.local.form_dict['back_civil']
        is_kuwaiti = frappe.local.form_dict['is_kuwaiti']

        # Load the Images

        front_image_path = upload_image(front_civil,"image1.png")
        back_image_path = upload_image(back_civil,"image2.png")

        front_text = get_front_side_civil_id_text(front_image_path, client, is_kuwaiti)
        back_text = get_back_side_civil_id_text(back_image_path, client, is_kuwaiti)

        result.update({'front_text': front_text})
        result.update({'back_text': back_text})
        
        return result
    
    except Exception as e:
        frappe.throw(e)

def get_front_side_civil_id_text(image_path, client, is_kuwaiti):
    """ This method fetches the image from the provided image path, calls the vision api to exrtract text from the image
        and parses through the obtained texts to get relevant texts for the required fields in the front side of the civil ID.

    Args:
        image_path (str): Path to image file
        client (obj): Vision API client library object
        is_kuwaiti (int): 0 for non-Kuwaiti and 1 for Kuwaiti

    Returns:
        dict: Dictionary of obtained texts for the required fields.
    """

    with io.open(image_path, 'rb') as image_file:
        content = image_file.read()

    image = vision.Image(content=content)

    response = client.text_detection(image=image)  # returns TextAnnotation

    texts = response.text_annotations
    
    result = {}
    assemble = {}
    index = 0
    print(texts[0].description)

    for index in range(1,len(texts)):
        assemble[index] = texts[index].description
    
    if is_kuwaiti == 1:
        result["Civil_ID_No"] = texts[find_index(assemble,"CARD")+1].description
        result["Country_Code"] = texts[find_index(assemble,"Nationality")+1].description
        result["Date_Of_Birth"] = datetime.datetime.strptime(texts[find_index(assemble,"Birth")+2].description, '%d/%m/%Y').strftime('%Y-%m-%d')
        result["Expiry_Date"] = datetime.datetime.strptime(texts[find_index(assemble,"Birth")+3].description, '%d/%m/%Y').strftime('%Y-%m-%d')
        if texts[find_index(assemble,"Sex")-1].description == "M" or texts[find_index(assemble,"Sex")-1].description == "F":
            result["Gender"] = texts[find_index(assemble,"Sex")-1].description
        else:
            result["Gender"] = ""

        result["Name"] = ""
        for i in range(find_index(assemble,"Name")+1,find_index(assemble,"Nationality")-2):
            result["Name"] = result["Name"] + texts[i].description + " "
        
        result["Arabic_Name"]= ""
        for i in range(find_index(assemble,"No")+1,find_index(assemble,"Name")-1):
            result["Arabic_Name"] = result["Arabic_Name"] + texts[i].description + " "

    else:
        result["Civil_ID_No"] = texts[find_index(assemble,"Civil")+3].description
        result["Country_Code"] = texts[find_index(assemble,"Nationality")+1].description
        result["Date_Of_Birth"] = datetime.datetime.strptime(texts[find_index(assemble,"لجن")+1].description, '%d/%m/%Y').strftime('%Y-%m-%d')
        result["Expiry_Date"] = datetime.datetime.strptime(texts[find_index(assemble,"لجن")+2].description, '%d/%m/%Y').strftime('%Y-%m-%d')
        result["Passport_Number"] = texts[find_index(assemble,"Nationality")+2].description
        result["Gender"] = ""
        if texts[find_index(assemble,"Sex")+1].description == "M" or texts[find_index(assemble,"Sex")+1].description == "F":
            result["Gender"] = texts[find_index(assemble,"Sex")+1].description

        result["Name"] = ""
        for i in range(find_index(assemble,"Name")+1,find_index(assemble,"Passport")):
            result["Name"] = result["Name"] + texts[i].description + " "
        
        result["Arabic_Name"]= ""
        for i in range(find_index(assemble,"الرقه")+1,find_index(assemble,"Name")):
            result["Arabic_Name"] = result["Arabic_Name"] + texts[i].description + " "
            
        result["Arabic_Name"] = result["Arabic_Name"][::-1]
    
    return result

def get_back_side_civil_id_text(image_path, client, is_kuwaiti):
    """ This method fetches the image from the provided image path, calls the vision api to exrtract text from the image
        and parses through the obtained texts to get relevant texts for the required fields in the back side side of the civil ID.

    Args:
        image_path (str): Path to image file
        client (obj): Vision API client library object
        is_kuwaiti (int): 0 for non-Kuwaiti and 1 for Kuwaiti

    Returns:
        dict: Dictionary of obtained texts for the required fields.
    """
    
    with io.open(image_path, 'rb') as image_file:
        content = image_file.read()

    image = vision.Image(content=content)

    response = client.text_detection(image=image)  # returns TextAnnotation

    texts = response.text_annotations
    
    result = {}
    assemble = {}
    index = 0
    
    for index in range(1,len(texts)):
        assemble[index] = texts[index].description
    if is_kuwaiti == 1:
        if find_index(assemble,"all"):
            result["PACI_No"] = texts[find_index(assemble,"all")-1].description
        else:
            result["PACI_No"] = " "
            
    else:
        result["PACI_No"] = ""
        if find_index(assemble,"YI"):
            result["PACI_No"] = texts[find_index(assemble,"YI")-1].description
        
        result["Sponsor_Name"]= ""
        if find_index(assemble, "(") and find_index(assemble, ")"):
            for i in range(find_index(assemble,")")+1,find_index(assemble,"العنوان:")):
                result["Sponsor_Name"] = result["Sponsor_Name"] + texts[i].description + " "
                
            result["Sponsor_Name"] = result["Sponsor_Name"][::-1]
    
    return result

@frappe.whitelist(allow_guest=True)
def get_passport_text():
    try:
        result = {}
        
        #initialize google vision client library
        os.environ['GOOGLE_APPLICATION_CREDENTIALS'] = cstr(frappe.local.site) + frappe.local.conf.google_application_credentials
        client = vision.ImageAnnotatorClient()
        
        front_passport = frappe.local.form_dict['front_passport']
        back_passport = frappe.local.form_dict['back_passport']

        front_image_path = upload_image(front_passport,hashlib.md5(front_passport.encode('utf-8')).hexdigest())
        back_image_path = upload_image(back_passport,hashlib.md5(back_passport.encode('utf-8')).hexdigest())

        front_text = get_passport_front_text(front_image_path, client)
        back_text = get_passport_back_text(back_image_path, client)
        
        result.update({'front_text': front_text})
        result.update({'back_text': back_text})
        
        return result
    
    except Exception as e:
        frappe.throw(e)

def get_passport_front_text(image_path, client):
    with io.open(image_path, 'rb') as image_file:
        content = image_file.read()

    image = vision.Image(content=content)

    response = client.text_detection(image=image)  # returns TextAnnotation

    texts = response.text_annotations
    
    result = {}
    assemble = {}
    index = 0
    
    for index in range(1,len(texts)):
        assemble[index] = texts[index].description
        
    text_length = len(texts)
    print(text_length)
    fuzzy = False
    if(text_length >= 5):
        if is_date(texts[text_length - 4].description, fuzzy):
            result["Passport_Date_of_Issue"] = datetime.datetime.strptime(texts[text_length - 4].description, '%d/%m/%Y').strftime('%Y-%m-%d')
        else:
            result["Passport_Date_of_Issue"] = ""
        if is_date(texts[text_length - 3].description, fuzzy):
            result["Passport_Date_of_Expiry"] = datetime.datetime.strptime(texts[text_length - 3].description, '%d/%m/%Y').strftime('%Y-%m-%d')
        else:
           result["Passport_Date_of_Expiry"] = "" 
        
        mrz = texts[text_length - 1].description
        
        result["Passport_Number"] = mrz.split("<")[0]
        
    return result

def get_passport_back_text(image_path, client):
    with io.open(image_path, 'rb') as image_file:
        content = image_file.read()

    image = vision.Image(content=content)

    response = client.text_detection(image=image)  # returns TextAnnotation

    texts = response.text_annotations
    
    result = {}
    assemble = {}
    index = 0
    
    for index in range(1,len(texts)):
        assemble[index] = texts[index].description
    
    result["Passport_Place_of_Issue"] = ""
    if find_index(assemble, "Place") and find_index(assemble, "of") and find_index(assemble, "Issue"):
        result["Passport_Place_of_Issue"] = texts[find_index(assemble, "Issue") + 3].description
        
    return result

def find_index(dictionary, word):
    for d in dictionary:
        if dictionary[d] == word:
            return d

def is_date(string, fuzzy=False):
    """
    Return whether the string can be interpreted as a date.

    :param string: str, string to check for date
    :param fuzzy: bool, ignore unknown tokens in string if True
    """
    try: 
        parse(string, fuzzy=fuzzy)
        return True

    except ValueError:
        return False

def upload_image(image, filename):
    """ This method writes a file to a server directory
    Args:
        image (str): Base64 encoded image
        filename (str): Name of the file
    Returns:
        str: Path to uploaded image
    """
    content = base64.b64decode(image)
    image_path = cstr(frappe.local.site)+"/private/files/user/"+filename

    with open(image_path, "wb") as fh:
        fh.write(content)
    
    return image_path

File: one_fm/templates/pages/applicant_docs.js
frappe.ready_events = [];
frappe.ready = function (fn) {
    frappe.ready_events.push(fn);
}
window.dev_server = {{dev_server}};
window.socketio_port = {{frappe.socketio_port}};

var is_kuwaiti = $('#First_Name').attr('is_kuwaiti');
var civil_id_reqd = $('#First_Name').attr('civil_id_reqd');

var applicant_name = $('#First_Name').attr('data');

const TOTAL_FORM_FIELDS = 22;

// List of fields relevant to CIVIL ID's front side text
const CIVIL_ID_FRONT_TEXT_FIELDS = [
  'First_Name',
  'Second_Name',
  'Third_Name',
  'Last_Name',
  'First_Arabic_Name',
  'Second_Arabic_Name',
  'Third_Arabic_Name',
  'Last_Arabic_Name',
  'Gender',
  'Civil_ID_No',
  'Country_Code',
  'Date_Of_Birth',
  'Expiry_Date',
]

// List of fields relevant to CIVIL ID's back side text
const CIVIL_ID_BACK_TEXT_FIELDS = [
  'PACI_No',
  'Sponsor_Name'
]

// List of fields relevant to passport front side text
const PASSPORT_FRONT_TEXT_FIELDS = [
  'Passport_Number',
  'Passport_Date_of_Issue',
  'Passport_Date_of_Expiry'
]

// List of fields relevant to passport back side text
const PASSPORT_BACK_TEXT_FIELDS = [
  'Passport_Place_of_Issue'
]

var front_cid_filepath = "";
var back_cid_filepath = "";
var front_passport_filepath = "";
var back_passport_filepath = "";

window.onload = () => {
  $(".required_indicator").hide();
  $("#tooltiptext1").hide();
  $("#tooltiptext2").hide();

  $("#Sponsor").hide();
  if(is_kuwaiti == 0){
    $("#Sponsor").show();
  }
  if(civil_id_reqd == 1){
    $(".required_indicator").show()
  }
  populate_nationality();
  populate_country();
}

let civil_id_image;
let passport_image;

function extract_image(){
  if (civil_id_reqd == 1){
    extract(document.getElementById("Civil_ID_Front").files[0],"Civil_ID","front_civil")
    extract(document.getElementById("Civil_ID_Back").files[0],"Civil_ID","back_civil")
  }
  extract(document.getElementById("Passport_Front").files[0],"Passport","front_passport")
  extract(document.getElementById("Passport_Back").files[0],"Passport","back_passport")
}

function upload_image(file, file_url, filename, token){
  var xhr = new XMLHttpRequest();
  let form_data = new FormData();

  xhr.open('POST', '/api/method/upload_file', true);

  xhr.setRequestHeader("X-Frappe-CSRF-Token", token);
  xhr.setRequestHeader("Accept", "application/json");

  form_data.append('file', file, filename);
  form_data.append('is_private', true)
  form_data.append('file_url', file_url)
  xhr.send(form_data);
}

function extract(file, type, key){

  if(file){
    let reader = new FileReader();

    reader.readAsDataURL(file);

    reader.onload = function() {
      result = reader.result;
      result = result.replace(/^data:image\/\w+;base64,/, "");
      if(type =="Civil_ID"){
        civil_id_image.append(key, result);
        return civil_id_image
      }
      else{
        passport_image.append(key, result);
        return passport_image
      }
    };
  }
}

function populate_nationality(){
  frappe.call({
    type: "GET",
    method: "one_fm.templates.pages.applicant_docs.populate_nationality",
    callback: function(r) {
      langArray = r.message;
      if(langArray){
        var nationality = document.getElementById("Nationality");
        for (let i=0; i<=langArray.length;i++) {
          nationality.options[nationality.options.length] = new Option(langArray[i], langArray[i]);
        }
      }
    }
  });
}
function populate_country(){
  frappe.call({
    type: "GET",
    method: "one_fm.templates.pages.applicant_docs.populate_country",
    callback: function(r) {
      langArray = r.message;

      if(langArray){
        var place_of_issue = document.getElementById("Passport_Place_of_Issue");
        var place_of_birth = document.getElementById("Birth_Place");
        for (let i=0; i<=langArray.length;i++) {
          place_of_issue.options[place_of_issue.options.length] = new Option(langArray[i], langArray[i]);
          place_of_birth.options[place_of_birth.options.length] = new Option(langArray[i], langArray[i]);
        }
      }
    }
  });
}

function fetchNationality(code){
  frappe.call({
    type: "GET",
    method: "one_fm.templates.pages.applicant_docs.fetch_nationality",
    args: {code :code},
    callback: function(r) {
      if(r.message){
        document.getElementById("Nationality").value = r.message;
      }
      else{
        document.getElementById("Nationality").value = "";
      }

    }
  });
}


function send_request(method, data, token, type){

  var request = new XMLHttpRequest();
  // POST to httpbin which returns the POST data as JSON
    request.open('POST', method ,true);
    request.setRequestHeader("X-Frappe-CSRF-Token", token );
    request.setRequestHeader("Accept", "application/json");

    request.send(data);
    request.type = type
    request.onreadystatechange = () => {
      if (request.readyState == XMLHttpRequest.DONE) {
        if (request.status === 200) {
          let r = null;
          try {
            r = JSON.parse(request.responseText);
            fill_form(r.message,request.type, token);
          } catch (e) {
            $("#cover-spin").hide();
            $('#finalForm').css('display', 'block');
            r = request.responseText;
          }

        } else if (request.status === 403) {
          $("#cover-spin").hide();
          $('#finalForm').css('display', 'block');
          let response = JSON.parse(request.responseText);
          frappe.msgprint({
            title: __("Not permitted"),
            indicator: "red",
            message: response._error_message,
          });
        } else {
          $("#cover-spin").hide();
          $('#finalForm').css('display', 'block');
          frappe.msgprint({
            title: __("Error extracting text"),
            indicator: "orange",
            message: __("Please fill out the below form manually."),
          });
        }
      }
    };
}


function upload(){
  // if($("#Civil_ID_Front").val().length == 0 ){
  //   $("#tooltiptext1").show()
  // }
  // else if($("#Civil_ID_Back").val().length == 0 ){
  //   $("#tooltiptext2").show();
  // }
  // else{
  civil_id_image = new FormData();
  passport_image = new FormData();
  extract_image();

  var method_map = {
    'civil_id': '/api/method/one_fm.templates.pages.applicant_docs.get_civil_id_text',
    'passport': '/api/method/one_fm.templates.pages.applicant_docs.get_passport_text'
  }
  frappe.call({
    type: "GET",
    method: "one_fm.templates.pages.applicant_docs.token",
    callback: function(r) {
      var token = r.message
       if (!!civil_id_image.entries().next().value){
        civil_id_image.append("is_kuwaiti",is_kuwaiti)
        $("#cover-spin").show(0);
        send_request(method_map['civil_id'], civil_id_image, token,"Civil ID")
      };
      if (!!passport_image.entries().next().value){
        $("#cover-spin").show(0);
        send_request(method_map['passport'], passport_image, token,"Passport")
        // disable button
        $('#fileUpload').prop('disabled', 1)
      };
    }
  });
  $('#declare-div').prop('hidden', true);
  $('#submitForm').prop('hidden', true);
}
// };


function fill_form(data, type,token){
  {/* This Function fills the output form for user to view.
  The value is fetched from the api*/}
  if(data == "Error"){
    alert("Sorry! Some Error Occured during uploading" + type);
  }
  else {

    let total_fill_counter = 0;

    if(type == "Civil ID"){
      input_data(data,'front_text','First_Name');
      input_data(data,'front_text','Second_Name');
      input_data(data,'front_text','Third_Name');
      input_data(data,'front_text','Last_Name');
      input_data(data,'front_text','First_Arabic_Name');
      input_data(data,'front_text','Second_Arabic_Name');
      input_data(data,'front_text','Third_Arabic_Name');
      input_data(data,'front_text','Last_Arabic_Name');
      input_data(data,'front_text','Gender');
      input_data(data,'front_text','Civil_ID_No');
      input_data(data,'front_text','Country_Code');
      input_data(data,'front_text','Date_Of_Birth');
      input_data(data,'front_text','Expiry_Date');
      input_data(data,'back_text','PACI_No');
      input_data(data,'front_text','Passport_Number');
      if(data['front_text']['Country_Code'] != undefined){
        fetchNationality(data['front_text']['Country_Code']);
      }

      front_cid_filepath = input_filepath(data, 'front_text', 'Civil_ID_Front',token)
      back_cid_filepath = input_filepath(data, 'back_text', 'Civil_ID_Back',token)

      if(is_kuwaiti==0){
        input_data(data,'back_text','Sponsor_Name');
      }

      let front_side_cid_filled_fields_count = count_filled_fields(data, 'front_text', CIVIL_ID_FRONT_TEXT_FIELDS);
      let back_side_cid_filled_fields_count = count_filled_fields(data, 'back_text', CIVIL_ID_BACK_TEXT_FIELDS);

      total_fill_counter += (front_side_cid_filled_fields_count + back_side_cid_filled_fields_count);
    }
    else if(type == "Passport"){
      // front_passport_filepath = input_filepath(data, 'front_text', 'Passport_Front',token)
      // back_passport_filepath = input_filepath(data, 'back_text', 'Passport_Back',token)

      function_set_passport_data(data)
      // input_data(data,'front_text','Passport_Number');
      // input_data(data,'front_text','Passport_Date_of_Issue');
      // input_data(data,'front_text','Passport_Date_of_Expiry');
      // input_data(data,'back_text','Passport_Place_of_Issue');

      // let front_side_pp_filled_fields_count = count_filled_fields(data, 'front_text', PASSPORT_FRONT_TEXT_FIELDS);
      // let back_side_pp_filled_fields_count = count_filled_fields(data, 'back_text', PASSPORT_BACK_TEXT_FIELDS);

      // total_fill_counter += (front_side_pp_filled_fields_count + back_side_pp_filled_fields_count);
    }

    // if (total_fill_counter < TOTAL_FORM_FIELDS){
    //   frappe.msgprint({
    //     title: __("Could not obtain all information"),
    //     indicator: "orange",
    //     message: __("Some fields in the below form may be empty. Please fill them out correctly."),
    //   });
    // }

  }
};

function count_filled_fields(data, text_side, list_of_keys){
  let fill_counter = 0;

  list_of_keys.forEach(field => {
    if(data[text_side][field]){
      fill_counter++;
    }
  });

  return fill_counter;
}

function input_filepath(Data, key1, key2,token){
  if(Data[key1][key2]!= undefined){
    upload_image(document.getElementById(key2).files[0],Data[key1][key2],applicant_name+"_"+key2+'.png',token)
    return Data[key1][key2]
  }
};

function sentenceCase (str) {
  if ((str===null) || (str===''))
    return false;
  else
  str = str.toString();

  return str.replace(/\w\S*/g,
  function(txt){return txt.charAt(0).toUpperCase() +
    txt.substr(1).toLowerCase();});
}


function function_set_passport_data(data){
  console.log(data);
  let doc = data.front_text;
  if(doc.surname){
    $('#Last_Name').val(doc.surname);
  }
  if (doc.birth_date){
    $('#Date_Of_Birth').val(doc.birth_date)
  }
  if (doc.gender){
    $('#Gender').val(doc.gender=='M' ? 'Male' : 'Female')
  }
  if (doc.id_number){
    $('#Passport_Number').val(doc.id_number)
  }
  if (doc.issuance_date){
    $('#Passport_Date_of_Issue').val(doc.issuance_date)
  }
  if (doc.expiry_date){
    $('#Passport_Date_of_Expiry').val(doc.expiry_date)
  }
  if (doc.birth_place){
    $.makeArray($('#Birth_Place>option')).forEach((item)=>{
      if (item.value.toLowerCase()==doc.birth_place.toLowerCase()){
        $('#Birth_Place').val(sentenceCase(doc.birth_place))
      }
    })
    
  }
  if (doc.given_names.length){
    let fields = ["#First_Name", "#Second_Name", "#Third_Name"]
    doc.given_names.forEach((item, index)=>{
      $(fields[index]).val(item);
    })
  }
  if($('#perdonal-detail :input')){
    $.makeArray($('#perdonal-detail :input')).forEach((item)=>{if (!item.value && item.required){$(`#${item.id}`).prop('style', 'border: 2px solid red;')};})
    $.makeArray($('select')).forEach((item)=>{if (!item.value && item.required && item.id){$(`#${item.id}`).prop('style', 'border: 2px solid red;')};})
  }

  // end
  $("#cover-spin").hide();
  $('#finalForm').css('display', 'block');
  get_uploaded_data(data);
  
  
  
}

function input_data(Data, key1, key2){
  if(Data[key1][key2]!= undefined){
    if(key2 =="Gender"){
      if(Data[key1][key2]=="M"){
        document.getElementById(key2).value = "Male"
      }
      else if(Data[key1][key2]=="F"){
        document.getElementById(key2).value = "Female"
      }
    }
    document.getElementById(key2).value = Data[key1][key2];
  }
}

function Submit(){
  var applicant_details = get_details_from_form();
  
  if($('#First_Name').attr("data")){
    // frappe.freeze();
		frappe.confirm('Are you sure you want to Submit?, On Submit the link will be expired!',
    () => {

      var frontPassport = document.getElementById('Passport_Front');
      var backPassport = document.getElementById('Passport_Back');

      upload_image_to_server(frontPassport.files[0]);
      upload_image_to_server(backPassport.files[0]);

      applicant_details['applicant_doc']['Passport Front'] = frontPassport.value;
      applicant_details['applicant_doc']['Passport Back'] =  backPassport.value;


			frappe.call({
				type: "POST",
				method: "one_fm.templates.pages.applicant_docs.update_job_applicant",
				args: {
					job_applicant: $('#First_Name').attr("data"),
					data: applicant_details
				},
				btn: this,
				callback: function(r){
					frappe.unfreeze();
					frappe.msgprint(frappe._("Succesfully Submitted your Details and our HR team will be responding to you soon."));
					if(r.message){
            setTimeout(()=>{window.location.href = "/careers"}, 3000);
					}
				}
			});
    }, () => {
      frappe.unfreeze();
        // action to perform if No is selected
    })
  }
  else{
    frappe.msgprint(frappe._("Please fill All the details to submit the Job Applicant"));
  }
}

function Save(){
  // $.makeArray($('#perdonal-detail :input')).forEach((item)=>{if (!item.value && item.required){$(`#${item.id}`).prop('style', 'border: 2px solid red;')};})
  // $.makeArray($('select')).forEach((item)=>{if (!item.value && item.required && item.id){$(`#${item.id}`).prop('style', 'border: 2px solid red;')};})
  let goodTogo = false;
    $.makeArray($('#perdonal-detail :input')).forEach((item)=>{if (!item.value && item.required){
      $('#declare-div').prop('hidden', true);
      $('#submitForm').prop('hidden', true);
      frappe.throw("Please fill all fields in red box.");
      return;
    };})

    $.makeArray($('select')).forEach((item)=>{if (!item.value && item.required){
      $('#declare-div').prop('hidden', true);
      $('#submitForm').prop('hidden', true);
      frappe.throw("Please fill all fields in red box.");
      return;
    };})

      var applicant_details = get_details_from_form();

      if($('#First_Name').attr("data")){
        // frappe.freeze();
        frappe.call({
          type: "POST",
          method: "one_fm.templates.pages.applicant_docs.save_as_draft",
          args: {
            job_applicant: $('#First_Name').attr("data"),
            data: applicant_details
          },
          btn: this,
          callback: function(r){
            frappe.unfreeze();
            frappe.msgprint(frappe._("Succesfully Saved your application as draft, you can finish up and Submit later!."));
            $('#declare-div').prop('hidden', false);
            $('#submitForm').prop('hidden', false);
          }
        });

      } else{
        frappe.msgprint(frappe._("Please, you need to fill some details before you can save "));
      }
  
}

function get_details_from_form() {
  var applicant_details = {};
  applicant_details['one_fm_first_name'] = $('#First_Name').val();
  applicant_details['one_fm_second_name'] = $('#Second_Name').val();
  applicant_details['one_fm_third_name'] = $('#Third_Name').val();
  applicant_details['one_fm_last_name'] = $('#Last_Name').val();
  applicant_details['one_fm_first_name_in_arabic'] = $('#First_Arabic_Name').val();
  applicant_details['one_fm_second_name_in_arabic'] = $('#Second_Arabic_Name').val();
  applicant_details['one_fm_third_name_in_arabic'] = $('#Third_Arabic_Name').val();
  applicant_details['one_fm_last_name_in_arabic'] = $('#Last_Arabic_Name').val();
  applicant_details['one_fm_gender'] = $('#Gender').val();
  applicant_details['one_fm_educational_qualification'] = $('#EducationalQualification').val();
  applicant_details['one_fm_university'] = $('#University').val();
  applicant_details['one_fm_marital_status'] = $('#MaritalStatus').val();
  applicant_details['one_fm_religion'] = $('#Religion').val();
  applicant_details['one_fm_date_of_birth'] = $('#Date_Of_Birth').val();
  applicant_details['one_fm_cid_number'] = $('#Civil_ID_No').val();
  applicant_details['one_fm_cid_expire'] = $('#Expiry_Date').val();
  applicant_details['one_fm_nationality'] = $('#Nationality').val();
  applicant_details['one_fm_place_of_birth'] = $('#Birth_Place').val();
  applicant_details['one_fm_passport_number'] = $('#Passport_Number').val();
  applicant_details['one_fm_passport_type'] = $('#Passport_Type').val();
  applicant_details['one_fm_passport_issued'] = $('#Passport_Date_of_Issue').val();
  applicant_details['one_fm_passport_expire'] = $('#Passport_Date_of_Expiry').val();
  applicant_details['one_fm_passport_holder_of'] = $('#Passport_Place_of_Issue').val();
  applicant_details['one_fm_country_code'] = $('#Country_Code').val();
  

  applicant_details['applicant_doc']={}

  
  get_filepath(applicant_details['applicant_doc'],front_cid_filepath, "Civil ID Front" )
  get_filepath(applicant_details['applicant_doc'],back_cid_filepath, "Civil ID Back" )
  //get_filepath(applicant_details['applicant_doc'],front_passport_filepath, "Passport Front" )
  //get_filepath(applicant_details['applicant_doc'],back_passport_filepath, "Passport Back" )

  // applicant_details['paci_no'] = $('#PACI_No').val();
  return applicant_details;
};

function get_filepath(object, value, key){
  if(value != ""){
    var file_name = key.replaceAll(" ","_")
    object[key] = applicant_name+"_"+file_name+'.png';
  }
  return object
}


function preview_front(input){
  var uploaded_pics = `
  <h4>Passport Front Image</h4>
  <img src="" alt="Front Page Of Passport" id="passport-front" class="form-control" style="height: 350px;">
`;

  document.getElementById("uploaded-passport-front").innerHTML = uploaded_pics;
  document.getElementById('passport-front').src = window.URL.createObjectURL(input.files[0]);
}


function preview_back(input){
  var uploaded_pics = `
  <h4>Passport Back Image</h4>
  <img src="" alt="Front Page Of Passport" id="passport-back" class="form-control" style="height: 350px;">
`;
  document.getElementById("uploaded-passport-back").innerHTML = uploaded_pics;
  document.getElementById('passport-back').src = window.URL.createObjectURL(input.files[0]);
}


function get_uploaded_data(data){
  return frappe.call({
    method: "one_fm.templates.pages.applicant_docs.get_uploaded_data",
    args: {
      data: data
    },
    callback: function(r){
      frappe.unfreeze();
      frappe.msgprint(frappe._(`The following were extracted from the Image <ul>${r.message.map(item => (
        `<li>${item}</li>`
      )).join('')}</ul> Kindly fill the remaining fields correctly.`));

    }
  })
}



function upload_image_to_server(file){
  var xhr = new XMLHttpRequest();
  let form_data = new FormData();

  xhr.open('POST', '/api/method/upload_file', false);

  xhr.setRequestHeader("X-Frappe-CSRF-Token", frappe.csrf_token);
  xhr.setRequestHeader("Accept", "application/json");

  form_data.append('file', file);
  form_data.append('is_private', true)
  xhr.send(form_data);

}
